source/common/router/scoped_rds.h

```diff
@@ -89,45 +91,67 @@ class ScopedRdsConfigSubscription : public Envoy::Config::DeltaConfigSubscriptio
+      envoy::api::v2::core::ConfigSource rds_config_source,
+      RouteConfigProviderManager& route_config_provider_manager,
       ScopedRoutesConfigProviderManager& config_provider_manager);
 
-  ~ScopedRdsConfigSubscription() override = default;
+  ~ScopedRdsConfigSubscription() override;
 
   const std::string& name() const { return name_; }
 
-  const ScopedConfigManager::ScopedRouteMap& scopedRouteMap() const {
-    return scoped_config_manager_.scopedRouteMap();
-  }
+  const ScopedRouteMap& scopedRouteMap() const { return scoped_route_map_; }
 
 private:
+  // Cleanup stale ScopedRouteInfo instances.
+  void cleanupDefferedDeletes();
+
   // Envoy::Config::DeltaConfigSubscriptionInstance
   void start() override { subscription_->start({}); }
 
   // Envoy::Config::SubscriptionCallbacks
+
+  // NOTE: state-of-the-world form onConfigUpdate(resources, version_info) will throw an
+  // EnvoyException on any error and essentially reject an update. While the Delta form
+  // onConfigUpdate(added_resources, removed_resources, version_info) by design will partially
+  // accept correct RouteConfiguration from management server.
   void onConfigUpdate(const Protobuf::RepeatedPtrField<ProtobufWkt::Any>& resources,
                       const std::string& version_info) override;
-  void onConfigUpdate(const Protobuf::RepeatedPtrField<envoy::api::v2::Resource>&,
-                      const Protobuf::RepeatedPtrField<std::string>&, const std::string&) override {
-    NOT_IMPLEMENTED_GCOVR_EXCL_LINE;
-  }
+  void onConfigUpdate(const Protobuf::RepeatedPtrField<envoy::api::v2::Resource>& added_resources,
+                      const Protobuf::RepeatedPtrField<std::string>& removed_resources,
+                      const std::string& version_info) override;
   void onConfigUpdateFailed(Envoy::Config::ConfigUpdateFailureReason,
                             const EnvoyException*) override {
-    ConfigSubscriptionCommonBase::onConfigUpdateFailed();
+    DeltaConfigSubscriptionInstance::onConfigUpdateFailed();
   }
   std::string resourceName(const ProtobufWkt::Any& resource) override {
     return MessageUtil::anyConvert<envoy::api::v2::ScopedRouteConfiguration>(resource,
                                                                              validation_visitor_)
         .name();
   }
 
+  // ScopedRouteInfo by scope name.
+  ScopedRouteMap scoped_route_map_;
+  // A map of (hash, scope-name), used to detect the key conflict between scopes.
+  absl::flat_hash_map<uint64_t, std::string> scope_name_by_hash_;
+  // For creating RDS subscriptions.
+  Server::Configuration::FactoryContext& factory_context_;
+  // Deferred to be deleted ScopedRouteInfo.
+  // As RdsRouteConfigProvider instance (which contains a tls slot) can only be destructed in main
+  // thread, we need to make sure the shared ScopedRouteInfo is deleted in main thread.
+  std::list<ScopedRouteInfoConstSharedPtr> deffered_to_be_deleted_;
```

Comment:
Why can't we directly just post to the main thread a cleanup lambda? Not a big fan of reimplementing deferred delete just for SRDS here.

Reply:
Remove this implementation now, as now the ownership of a RdsRouteConfigProivder is kept within SRDS subscription.